ecs-agent/acs/session/session_test.go

```diff
@@ -0,0 +1,1328 @@
+	discoverEndpointClient.EXPECT().DiscoverPollEndpoint(testconst.ContainerInstanceARN).Return(fakeServer.URL, nil)
+
+	ended := make(chan bool, 1)
+	go func() {
+		acsSession := NewSession(testconst.ContainerInstanceARN,
+			testconst.ClusterName,
+			discoverEndpointClient,
+			testCreds,
+			noopFunc,
+			acsclient.NewACSClientFactory(),
+			metricsfactory.NewNopEntryFactory(),
+			agentVersion,
+			agentGitShortHash,
+			"containerd",
+			testMinAgentConfig,
+			nil,
+			nil,
+			nil,
+			nil,
+			nil,
+			nil,
+			nil,
+			nil,
+			nil,
+			resourceHandler,
+			nil,
+		)
+
+		acsSession.Start(ctx)
+		// Start should never return unless the context is canceled.
+		ended <- true
+	}()
+
+	// WaitGroup is necessary to wait for HandleResourceAttachment to be called in separate goroutine before exiting
+	// the test.
+	wg := sync.WaitGroup{}
+	wg.Add(1)
+	resourceHandler.EXPECT().HandleResourceAttachment(gomock.Any()).Do(func(arg0 interface{}) {
+		defer wg.Done() // decrement WaitGroup counter now that HandleResourceAttachment function has been called
+	})
+
+	serverIn <- sampleAttachResourceMessage
+
+	select {
+	case err := <-errChan:
+		t.Fatal("Error should not have been returned from server", err)
+	case <-ctx.Done():
+		// Context is canceled when requestsChan receives an ACK.
+	}
+
+	wg.Wait()
+
+	fakeServer.Close()
+	// Cancel context should close the session.
+	<-ended
+}
+
+// TestSessionCallsAddUpdateRequestHandlers tests that the Session calls the function contained in its struct field
+// 'addUpdateRequestHandlers' is called if it is not nil and the Session's dockerVersion is not "containerd".
+func TestSessionCallsAddUpdateRequestHandlers(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	addUpdateRequestHandlersCalled := false
+	addUpdateRequestHandlers := func(cs wsclient.ClientServer) {
+		addUpdateRequestHandlersCalled = true
+	}
+
+	discoverEndpointClient := mock_api.NewMockECSDiscoverEndpointSDK(ctrl)
+	discoverEndpointClient.EXPECT().DiscoverPollEndpoint(gomock.Any()).Return(acsURL, nil).AnyTimes()
+	ctx, cancel := context.WithCancel(context.Background())
+
+	mockWsClient := mock_wsclient.NewMockClientServer(ctrl)
+	mockClientFactory := mock_wsclient.NewMockClientFactory(ctrl)
+	mockClientFactory.EXPECT().
+		New(gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any(), gomock.Any()).
+		Return(mockWsClient).AnyTimes()
+	mockWsClient.EXPECT().SetAnyRequestHandler(gomock.Any()).AnyTimes()
+	mockWsClient.EXPECT().AddRequestHandler(gomock.Any()).AnyTimes()
+	mockWsClient.EXPECT().Connect(gomock.Any(), gomock.Any(), gomock.Any()).
+		Return(time.NewTimer(wsclient.DisconnectTimeout), nil).AnyTimes()
+	mockWsClient.EXPECT().Serve(gomock.Any()).Do(func(interface{}) {
+		if addUpdateRequestHandlersCalled {
+			cancel()
+		}
+	})
+	mockWsClient.EXPECT().WriteCloseMessage().Return(nil).AnyTimes()
+	mockWsClient.EXPECT().Close().Return(nil).AnyTimes()
```

Comment:
I see more or less this batch of `EXPECT()`s in most tests, could/should they be moved to another method?

Reply:
They are not the exact same across every test and can differ depending on the test. Since there are too many differences to take into account depending on the test, IMO writing such a method would add more complexity than simplicity.

Reply:
ack, this may be a good idea if adding more tests grows cumbersome later down the road